evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,551 @@
+        virtual_first_change,
+    )
+}
+
+pub fn generate_range_check_value<F: RichField>(
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+    timestamp: &[F],
+    context_first_change: &[F],
+    segment_first_change: &[F],
+    virtual_first_change: &[F],
+) -> Vec<F> {
+    let num_ops = context.len();
+    let mut range_check = Vec::new();
+
+    for idx in 0..num_ops - 1 {
+        let this_address_unchanged = F::ONE
+            - context_first_change[idx]
+            - segment_first_change[idx]
+            - virtual_first_change[idx];
+
+        range_check.push(
+            context_first_change[idx] * (context[idx + 1] - context[idx] - F::ONE)
+                + segment_first_change[idx] * (segment[idx + 1] - segment[idx] - F::ONE)
+                + virtual_first_change[idx] * (virtuals[idx + 1] - virtuals[idx] - F::ONE)
+                + this_address_unchanged * (timestamp[idx + 1] - timestamp[idx] - F::ONE),
+        );
+    }
+
+    range_check.push(F::ZERO);
+
+    range_check
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> MemoryStark<F, D> {
+    pub(crate) fn generate_trace_rows(
+        &self,
+        memory_ops: Vec<(F, F, F, F, F, [F; 8])>,
+    ) -> Vec<[F; NUM_REGISTERS]> {
+        let num_ops = memory_ops.len();
+
+        let mut trace_cols: [Vec<F>; NUM_REGISTERS] = vec![vec![F::ZERO; num_ops]; NUM_REGISTERS]
+            .try_into()
+            .unwrap();
+        for i in 0..num_ops {
+            let (timestamp, is_read, context, segment, virt, values) = memory_ops[i];
+            trace_cols[TIMESTAMP][i] = timestamp;
+            trace_cols[IS_READ][i] = is_read;
+            trace_cols[ADDR_CONTEXT][i] = context;
+            trace_cols[ADDR_SEGMENT][i] = segment;
+            trace_cols[ADDR_VIRTUAL][i] = virt;
+            for j in 0..8 {
+                trace_cols[value_limb(j)][i] = values[j];
+            }
+        }
+
+        self.generate_memory(&mut trace_cols);
+
+        let mut trace_rows = vec![[F::ZERO; NUM_REGISTERS]; num_ops];
+        for (i, col) in trace_cols.iter().enumerate() {
+            for (j, &val) in col.iter().enumerate() {
+                trace_rows[j][i] = val;
+            }
+        }
+        trace_rows
+    }
+
+    fn generate_memory(&self, trace_cols: &mut [Vec<F>]) {
+        let num_trace_rows = trace_cols[0].len();
+
+        let timestamp = &trace_cols[TIMESTAMP];
+        let is_read = &trace_cols[IS_READ];
+        let context = &trace_cols[ADDR_CONTEXT];
+        let segment = &trace_cols[ADDR_SEGMENT];
+        let virtuals = &trace_cols[ADDR_VIRTUAL];
+        let values: Vec<[F; 8]> = (0..num_trace_rows)
+            .map(|i| {
+                let arr: [F; 8] = (0..8)
+                    .map(|j| &trace_cols[value_limb(j)][i])
+                    .cloned()
+                    .collect_vec()
+                    .try_into()
+                    .unwrap();
+                arr
+            })
+            .collect();
```

Comment:
Transpose doesn't seem necessary since the transpose is taken again below. So we could just have `values: [Vec<F>, 8]`.

Reply:
Hm I think it's necessary for it to be by row, so we can sort it in `sort_memory_ops`?